File: backend/app/modules/estoque/schemas/movimentacao.py
from pydantic import BaseModel, Field
from typing import Optional
from datetime import datetime
from decimal import Decimal
from enum import Enum


class TipoMovimentacao(str, Enum):
    ENTRADA = "ENTRADA"
    SAIDA = "SAIDA"
    TRANSFERENCIA = "TRANSFERENCIA"
    AJUSTE_ENTRADA = "AJUSTE_ENTRADA"
    AJUSTE_SAIDA = "AJUSTE_SAIDA"
    INVENTARIO = "INVENTARIO"
    PRODUCAO = "PRODUCAO"
    DEVOLUCAO = "DEVOLUCAO"


# Schemas para MovimentacaoEstoque
class MovimentacaoBase(BaseModel):
    tipo: TipoMovimentacao = Field(..., description="Tipo de movimentação")
    item_id: int = Field(..., description="ID do item")
    quantidade: Decimal = Field(..., gt=0, description="Quantidade movimentada")
    unidade_id: int = Field(..., description="ID da unidade de medida")
    lote_id: Optional[int] = Field(None, description="ID do lote (opcional)")
    local_id: Optional[int] = Field(None, description="ID do local da movimentação")
    numero: Optional[str] = Field(None, max_length=50, description="Número do documento")
    serie: Optional[str] = Field(None, max_length=10, description="Série do documento")
    observacoes: Optional[str] = Field(None, description="Observações")
    custo_unitario: Optional[Decimal] = Field(None, description="Custo unitário (para entradas)")
    usuario: Optional[str] = Field(None, max_length=100, description="Usuário responsável")


class MovimentacaoCreate(MovimentacaoBase):
    data_movimentacao: Optional[datetime] = Field(None, description="Data da movimentação (se não informada, usa a data atual)")


class MovimentacaoUpdate(BaseModel):
    observacoes: Optional[str] = None
    numero: Optional[str] = None
    serie: Optional[str] = None


class Movimentacao(MovimentacaoBase):
    id: int
    data_movimentacao: datetime
    saldo_anterior: Optional[Decimal] = None
    saldo_atual: Optional[Decimal] = None
    item_codigo: Optional[str] = None
    item_nome: Optional[str] = None
    unidade_sigla: Optional[str] = None
    lote_codigo: Optional[str] = None
    local_nome: Optional[str] = None
    created_at: datetime
    updated_at: Optional[datetime] = None

    class Config:
        from_attributes = True
        json_encoders = {
            Decimal: lambda v: str(v).replace('.', ',') if v is not None else None
        }


# Schema para consulta com dados completos
class MovimentacaoDetalhada(Movimentacao):
    item_codigo: Optional[str] = None
    item_descricao: Optional[str] = None
    unidade_sigla: Optional[str] = None
    lote_codigo: Optional[str] = None
    # Mantido apenas um local

File: frontend/features/estoque/types/movimentacao.ts
export enum TipoMovimentacao {
  ENTRADA = 'ENTRADA',
  SAIDA = 'SAIDA',
  TRANSFERENCIA = 'TRANSFERENCIA',
  AJUSTE_ENTRADA = 'AJUSTE_ENTRADA',
  AJUSTE_SAIDA = 'AJUSTE_SAIDA',
  INVENTARIO = 'INVENTARIO',
  PRODUCAO = 'PRODUCAO',
  DEVOLUCAO = 'DEVOLUCAO',
}

export interface MovimentacaoEstoque {
  id: number;
  tipo: string;
  item_id: number;
  item_codigo?: string;
  item_nome?: string;
  quantidade: number;
  unidade_id: number;
  unidade_sigla?: string;
  lote_id?: number;
  lote_codigo?: string;
  local_id?: number;
  local_nome?: string;
  data_movimentacao: string;
  numero?: string;
  serie?: string;
  observacoes?: string;
  custo_unitario?: number;
  saldo_anterior?: number;
  saldo_atual?: number;
  usuario?: string;
  created_at: string;
  updated_at?: string;
}

export interface CreateMovimentacaoEntradaDTO {
  item_id: number;
  unidade_id: number;
  local_id: number;
  lote_id?: number;
  quantidade: number;
  custo_unitario: number;
  data_movimentacao?: string;
  numero?: string;
  serie?: string;
  observacoes?: string;
  usuario?: string;
}

export interface CreateMovimentacaoSaidaDTO {
  item_id: number;
  unidade_id: number;
  local_id: number;
  lote_id?: number;
  quantidade: number;
  custo_unitario?: number;
  data_movimentacao?: string;
  numero?: string;
  serie?: string;
  observacoes?: string;
  usuario?: string;
}

// Transferência removida do backend (mantido para compatibilidade caso necessário)
export interface CreateMovimentacaoTransferenciaDTO {
  item_id: number;
  unidade_id: number;
  local_id: number; // Representa o local da movimentação
  lote_id?: number;
  quantidade: number;
  custo_unitario?: number;
  data_movimentacao?: string;
  numero?: string;
  serie?: string;
  observacoes?: string;
  usuario?: string;
}

export interface CreateMovimentacaoAjusteDTO {
  tipo: 'AJUSTE_ENTRADA' | 'AJUSTE_SAIDA';
  item_id: number;
  unidade_id: number;
  local_id: number;
  lote_id?: number;
  quantidade: number;
  custo_unitario?: number;
  data_movimentacao?: string;
  numero?: string;
  serie?: string;
  observacoes?: string;
  usuario?: string;
}

export interface MovimentacaoFilters {
  item_id?: number;
  local_id?: number;
  data_inicio?: string;
  data_fim?: string;
  tipo_movimentacao?: TipoMovimentacao;
}
